panther-labs/panther
# Managed s3 bucket notifications

## Background

Enable Panther to manage bucket notifications for the user. If user allows, Panther will create an SNS topic in the bucket's region and configure the bucket to send SNS notifications to that topic.

## Changes

- Add field `managedBucketNotifications` field in the source api so that user can select whether they want Panther to manage their bucket notifications configuration
- Add field `notificationsConfigurationSucceeded` so that the frontend knows whether this operation succeeded or not. Even if managing bucket notifications fail, Panther proceeds to save the source. The user just have to set them up manually.
- When an s3 source is created/updated and `managedBucketNotifications` is set, Panther sync's the updates the bucket's notification configuration.A notification configuration is created for every prefix the user has specified in the s3 source.
- When an s3 source is deleted, Panther deletes the Panther-managed notifications from the bucket. Any user-created configuration is left intact. Panther does not delete the SNS topic created because it may have been re-used. The UI will show the user that this resource has not been deleted.

## Testing

- Unit
- Manual e2e - All tests are done with a bucket from another account (Panther in @giorgosp account, bucket in @kostaspap account)
    - Create source with managed notifications with some prefixes (prefixa, prefixb) 
        - Verify bucket notifications config
    - Add some bucket notifications manually (not only SNS but also a Lambda notification). Update source prefixes. 
        - Verify bucket notifications config
    - Update source with a prefix already included in the bucket 
        - Verify error response (source is updated, but bucket notifications could not be configured. AWS doesn't allow overlapping prefixes)
    - Upload a file to the bucket - Verified event ended up in Athena
    - Delete source - Only Panther-managed configs should be removed from the bucket
    -  Create a second source for a second bucket in the same region as the first. Verify bucket notifications configured correctly and event made it to our data lake.
